qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,213 @@
+
+    In this case the evaluation of the function has a chance of failing. The user, with specific knowledge about his function can catch this errors and handle before
+    passing the result to the optimizer.
+
+    In case the user isn't dealing with complicated function and is more familiar with step by step optimization algorithms, a method step() has been created that
+    acts as a wrapper for ask() and tell().
+    In the same spirit the method minimize() will optimize the function and return the result without the user having to worry about the optimization process.
+
+    To see other libraries that use this interface one can visit: https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html
+
+
+    """
+
+    def __init__(
+        self,
+        callback: Optional[CALLBACK] = None,
+        maxiter: int = None,
+
+    ):
+        """
+        Args:
+            maxiter: Number of steps in the optimization process before ending the loop.
+        """
+        super().__init__()
+        self._state = None
+        self.callback = callback
+        self.maxiter = maxiter  # Remove maxiter
+
+    def ask(self) -> AskObject:
+        """
+        Returns the information on how to evaluate the objective function.
+        Canonical optimization workflow using ask() and tell() can be seen in SteppableOptimizer.step().
+        Returns:
+            AskOjbect: Since the way to evaluate the function can vary much with different optimization algorithms, the object will be
+        a custom dataclass for each optimizer.
+        """
+        raise NotImplementedError
+
+    def tell(self, ask_object: AskObject, tell_object: TellObject) -> None:
+        """
+        Updates the optimization state once the objective function has been evaluated.
+        Canonical optimization workflow using ask() and tell() can be seen in SteppableOptimizer.step().
+        Args:
+            ask_object: Contains the information on how the evaluation was done.
+            tell_object: Contains all relevant information about the evaluation of the objective function.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def evaluate(self, ask_object: AskObject) -> TellObject:
+        """
+        Evaluates the function according to the instructions contained in ask_object.
+        If the user decides to use step() instead of ask() and tell() this function will contain the logic on how to evaluate
+        the function.
+        Args:
+            ask_object: Contains the information on how to do the evaluation.
+        Return:
+            tell_object: Contains all relevant information about the evaluation of the objective function.
+        """
+        raise NotImplementedError
+
+    def user_evaluate(
+        self, *args, **kwargs
+    ) -> TellObject:  # Maybe this function should be abstract. For some optimizers (The return type with a custom tell object the function needs to be overwriten)
+        """
```

Comment:
The doc needs to explain why this is needed.

Reply:
From our discussion I think this method can be removed entirely 🙂